### ISSUE TITLE
undesirable_function_linter could be namespace-aware

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#1770]
Currently we ignore namespaces as all our defaults are `base` exports, but downstreams could be dealing with more ambiguous situations where namespace-qualified lints are better, e.g. `run_model()` is exported by `foo` and `bar`, but only `bar::run_model()` is undesirable:

```r
undesirable_function_linter(c(`bar::run_model` = NA))
```

Should "just work" and we add a `SYMBOL_PACKAGE` requirement on the XPath for `bar::model` and any other elements matching `::`.

**Comment by AshesITR** [comment]
Is there something that could be done for `library(bar)` followed by unqualified usage?

Maybe make an undesirable_library_linter?

I realize that we shouldn't do this in undesirable_function_linter because of false positives due to export masking like

```
library(bar)
library(foo)
run_model()
```

**Comment by MichaelChirico** [comment]
For the use I have in mind, I would avoid complicating things by ignoring `run_model()` usage and only linting `bar::run_model()`. There's some dragons down that route -- presumably we'd want to ensure that `run_model` is resolving to `bar::run_model()` instead of `foo::run_model()`, which involves reconstructing the search path from the recursive DESCRIPTION files, and even then it's an approximation because of `source()`/`read_chunk()`/etc... messy.

**Comment by AshesITR** [comment]
Completely agreed. I was thinking instead of solving the halting problem we could make it possible to lint `library(bar)` calls in a separate linter.

**Comment by MichaelChirico** [comment]
Got it... noting that it doesn't _quite_ apply in this case (since we aren't necessarily extending to say `bar::great_function()` is also linted). But there are cases (e.g. reading JSON files, for which there are at least 3 packages I can think of immediately) where maintainers might want to align on a single choice and discourage alternatives.
